building-epics/posix-threads-priority-scheduling-linux.md

```diff
@@ -0,0 +1,145 @@
+
+If you are starting IOCs using user-mode systemd (i.e. with `systemctl --user` commands) you will also need the above settings or an equivalent for the user account(s) that you run your IOC under.
+
+### IOCs as systemd Services
+
+On Linux systems using systemd you probably already have a service file in `/etc/systemd/system/myioc.service` or elsewhere that starts your IOC as a service. The following example uses procServ and includes the required configuration for granting the IOC process the necessary permissions.
+
+``` systemd
+[Unit]
+Description=EPICS Soft IOC myioc
+Requires=network.target
+After=network.target
+
+[Service]
+ExecStart=/usr/bin/procServ --foreground --quiet --port 4051 /path_to_myioc/iocBoot/myiocinstance/st.cmd
+Restart=always
+User=iocuser
+RuntimeDirectory=myioc
+CPUAffinity=1-3
+LimitMEMLOCK=infinity
+LimitRTPRIO=99
+
+[Install]
+WantedBy=multi-user.target
+```
+
+Note the last three lines of the `[Service]` section which are granting the process permissions to run in real-time. Refer to the [systemd.exec documentation](https://www.freedesktop.org/software/systemd/man/systemd.exec.html#Process%20Properties) for details.
+
+**Tip:** In case your service file is generated automatically by a deployment tool which doesn't allow you to configure these real-time parameters you can augment an existing service file by dropping the real-time related configuration into `/etc/systemd/system/myioc.service.d/10-realtime.conf`:
+
+``` systemd
+[Service]
+CPUAffinity=1-3
+LimitMEMLOCK=infinity
+LimitRTPRIO=99
+```
+
+Once your service file is in place, instruct systemd to reload its configuration and to start your IOC:
+
+``` console
+$ sudo systemctl daemon-reload
+$ sudo systemctl start myioc.service
+```
+
+### Summary
+
+You need to do three things:
+
+1.  Build EPICS base with `USE_POSIX_THREAD_PRIORITY_SCHEDULING=YES`
+2.  Make sure your EPICS application process has sufficient privileges to use SCHED_FIFO and the desired priority range.
+3.  _SysV Init only:_ Set the soft and hard RTPRIO limits for your EPICS application process to enable it to actually make use of the privilege. This step is required because it is usually a bad idea to run _all_ processes created by a given user or group at real-time priorities.  
+    _Systemd only:_ Set the permissions for RTPRIO and MEMLOCK in the `[Service]` section of your IOC service configuration.
+
+## Verify that Real-Time Scheduling has been enabled correctly
+
+### Verify that Memory has been locked
+
+Run
+
+``` console
+$ grep Vm /proc/`pgrep st.cmd`/status
```

Comment:
I would prefer to mention that this would work only if there is only one IOC running on the host, this way we keep the original idea of the how to to keep simple (good for newcomers)